README.md
# winbgim64
WinBGIm for 64-bit Windows

This is a port of the WinBGIm library (and graphics.h) for 64-bit systems. The library file generated from building it is called libbgi64.a.


WinBGIm64/changelog.txt
Changelog:

winbgim.cxx 302: change duplicate int right to int top (2nd arg)

winthread.cxx 103: SetWindowLong->SetWindowLongPtr (deprecated in Win64), GWL_USERDATA->GWLP_USERDATA (deprecated in Win64), (LONG)pWndData->(LONG_PTR)pWndData (deprecated in Win64)

drawing.cxx 66: GetWindowLong->GetWindowLongPtr (deprecated in Win64), GWL_USERDATA->GWLP_USERDATA (deprecated in Win64)

winbgim.cxx 464: "EGAVGA"->(char *) "EGAVGA" (deprecated in later versions of C++)

winbgim.cxx 570: cast all string constants to (char *) (deprecated in later versions of C++)

text.cxx 71: cast all string constants to (char *) (deprecated in later versions of C++)

text.cxx 354: "X"->(char *) "X" (deprecated in later versions of C++)

*The makefile can only be run on Unix systems, but it is possible to build the library from scratch on Windows systems using 64-bit minGW!

WinBGIm64/src/text.cxx
// File: text.cpp
// Written by:
//      Grant Macklem (Grant.Macklem@colorado.edu)
//      Gregory Schmelter (Gregory.Schmelter@colorado.edu)
//      Alan Schmidt (Alan.Schmidt@colorado.edu)
//      Ivan Stashak (Ivan.Stashak@colorado.edu)
// CSCI 4830/7818: API Programming
// University of Colorado at Boulder, Spring 2003
// http://www.cs.colorado.edu/~main/bgi
//
// This file contains the code necessary to draw/modify text
//

#include <windows.h>        // Provides the Win32 API
#include <windowsx.h>       // Provides GDI helper macros
#include <iostream>
#include <sstream>          // Provides ostringstream
#include <string>           // Provides string
#include "winbgim.h"         // API routines
#include "winbgitypes.h"    // Internal structure data


/*****************************************************************************
*
*   Some very useful arrays -- Same as previous version for consistency
*   Also, the exported definition of bgiout.
*
*****************************************************************************/
std::ostringstream bgiout;
static int font_weight[] =
{
    FW_BOLD,    // DefaultFont
    FW_NORMAL,  // TriplexFont
    FW_NORMAL,  // SmallFont
    FW_NORMAL,  // SansSerifFont
    FW_NORMAL,  // GothicFont
    FW_NORMAL,  // ScriptFont
    FW_NORMAL,  // SimplexFont
    FW_NORMAL,  // TriplexScriptFont
    FW_NORMAL,  // ComplexFont
    FW_NORMAL,  // EuropeanFont
    FW_BOLD     // BoldFont
};
static int font_family[] =
{
    FIXED_PITCH|FF_DONTCARE,     // DefaultFont
    VARIABLE_PITCH|FF_ROMAN,     // TriplexFont
    VARIABLE_PITCH|FF_MODERN,    // SmallFont
    VARIABLE_PITCH|FF_DONTCARE,  // SansSerifFont
    VARIABLE_PITCH|FF_SWISS,     // GothicFont
    VARIABLE_PITCH|FF_SCRIPT,    // ScriptFont
    VARIABLE_PITCH|FF_DONTCARE,  // SimplexFont
    VARIABLE_PITCH|FF_SCRIPT,    // TriplexScriptFont
    VARIABLE_PITCH|FF_DONTCARE,  // ComplexFont
    VARIABLE_PITCH|FF_DONTCARE,  // EuropeanFont
    VARIABLE_PITCH|FF_DONTCARE   // BoldFont
};
static char* font_name[] =
{
    (char *) "Console",          // DefaultFont
    (char *) "Times New Roman",  // TriplexFont
    (char *) "Small Fonts",      // SmallFont
    (char *) "MS Sans Serif",    // SansSerifFont
    (char *) "Arial",            // GothicFont
    (char *) "Script",           // ScriptFont
    (char *) "Times New Roman",  // SimplexFont
    (char *) "Script",           // TriplexScriptFont
    (char *) "Courier New",      // ComplexFont
    (char *) "Times New Roman",  // EuropeanFont
    (char *) "Courier New Bold", // BoldFont
};

static struct { int width; int height; } font_metrics[][11] = {
    {{0,0},{8,8},{16,16},{24,24},{32,32},{40,40},{48,48},{56,56},{64,64},{72,72},{80,80}}, // DefaultFont
    {{0,0},{13,18},{14,20},{16,23},{22,31},{29,41},{36,51},{44,62},{55,77},{66,93},{88,124}}, // TriplexFont
    {{0,0},{3,5},{4,6},{4,6},{6,9},{8,12},{10,15},{12,18},{15,22},{18,27},{24,36}}, // SmallFont
    {{0,0},{11,19},{12,21},{14,24},{19,32},{25,42},{31,53},{38,64},{47,80},{57,96},{76,128}}, // SansSerifFont
    {{0,0},{13,19},{14,21},{16,24},{22,32},{29,42},{36,53},{44,64},{55,80},{66,96},{88,128}}, // GothicFont

    // These may not be 100% correct
    {{0,0},{11,19},{12,21},{14,24},{19,32},{25,42},{31,53},{38,64},{47,80},{57,96},{76,128}}, // ScriptFont
    {{0,0},{11,19},{12,21},{14,24},{19,32},{25,42},{31,53},{38,64},{47,80},{57,96},{76,128}}, // SimplexFont
    {{0,0},{13,18},{14,20},{16,23},{22,31},{29,41},{36,51},{44,62},{55,77},{66,93},{88,124}}, // TriplexScriptFont
    {{0,0},{11,19},{12,21},{14,24},{19,32},{25,42},{31,53},{38,64},{47,80},{57,96},{76,128}}, // ComplexFont
    {{0,0},{11,19},{12,21},{14,24},{19,32},{25,42},{31,53},{38,64},{47,80},{57,96},{76,128}}, // EuropeanFont
    {{0,0},{11,19},{12,21},{14,24},{19,32},{25,42},{31,53},{38,64},{47,80},{57,96},{76,128}} // BoldFont
};

// horiz     LEFT_TEXT   0  left-justify text  
//           CENTER_TEXT 1  center text  
//           RIGHT_TEXT  2  right-justify text  
// vertical  BOTTOM_TEXT 0  bottom-justify text  
//           CENTER_TEXT 1  center text  
//           TOP_TEXT    2  top-justify text 
static UINT horiz_align[3] = {TA_LEFT, TA_CENTER, TA_RIGHT};
static UINT vert_align[3] = {TA_BOTTOM, TA_BASELINE, TA_TOP};

/*****************************************************************************
*
*   Some helper functions
*
*****************************************************************************/

// This function 
// POSTCONDITION:
//
static void set_align(WindowData* pWndData)
{
    UINT alignment = pWndData->alignment;

    // if we are vertical, things are swapped.
    if (pWndData->textInfo.direction == HORIZ_DIR)
    {
	alignment |= horiz_align[pWndData->textInfo.horiz];
	alignment |= vert_align[pWndData->textInfo.vert];
    }
    else
    {
	alignment |= horiz_align[pWndData->textInfo.vert];
	alignment |= vert_align[pWndData->textInfo.horiz];
    }

    // set the alignment for all valid pages.
    for ( int i = 0; i < MAX_PAGES; i++ )
	SetTextAlign(pWndData->hDC[i], alignment);
}

// This function updates the current hdc with the user defined font
// POSTCONDITION: text written to the current hdc will be in the new font
//
static void set_font(WindowData* pWndData)
{
    int mindex;
    double xscale, yscale;
    HFONT hFont;
    
    // get the scaling factors based on charsize
    if(pWndData->textInfo.charsize == 0)
    {
	xscale = pWndData->t_scale[0] / pWndData->t_scale[1];
	yscale = pWndData->t_scale[2] / pWndData->t_scale[3];
		
	// if font zero, only use factors.. else also multiply by 4
	if (pWndData->textInfo.font == 0)
	    mindex = 0;
	else
	    mindex = 4;
    }
    else
    {
	xscale = 1.0;
	yscale = 1.0;
	mindex = pWndData->textInfo.charsize;
    }

    // with the scaling decided, make a font.
    hFont = CreateFont(
	int(font_metrics[pWndData->textInfo.font][mindex].height * yscale), 
	int(font_metrics[pWndData->textInfo.font][mindex].width  * xscale),
	pWndData->textInfo.direction * 900,
	(pWndData->textInfo.direction & 1) * 900,
	font_weight[pWndData->textInfo.font],
	FALSE,
	FALSE,
	FALSE,
	DEFAULT_CHARSET,
	OUT_DEFAULT_PRECIS,
	CLIP_DEFAULT_PRECIS,
	DEFAULT_QUALITY,
	font_family[pWndData->textInfo.font],
	font_name[pWndData->textInfo.font]
	);

    // assign the fonts to each of the hdcs
    for ( int i = 0; i < MAX_PAGES; i++ )
	SelectObject( pWndData->hDC[i], hFont );
}


/*****************************************************************************
*
*   The actual API calls are implemented below
*
*****************************************************************************/
// This function fills the textsettingstype structure pointed to by textinfo
// with information about the current text font, direction, size, and
// justification.
// POSTCONDITION: texttypeinfo has been filled with the proper information
//
void gettextsettings(struct textsettingstype *texttypeinfo)
{
    // if its null, leave.
    if (!texttypeinfo)
	return;
	
    WindowData* pWndData = BGI__GetWindowDataPtr( );

    *texttypeinfo = pWndData->textInfo;	
}

// This function prints textstring to the screen at the current position
// POSTCONDITION: text has been written to the screen using the current font,
//                direction, and size.  In addition, the current position has
//                been modified to reflect the text that was just output.
//
void outtext(char *textstring)
{
    HDC hDC = BGI__GetWinbgiDC( );
    WindowData* pWndData = BGI__GetWindowDataPtr( );

    // so we can clear the screen
    POINT cp;
    GetCurrentPositionEx( hDC, &cp );

    // check cp alignment
    if (pWndData->alignment != TA_UPDATECP)
    {
	pWndData->alignment = TA_UPDATECP;
	set_align(pWndData);
    }

    TextOut(hDC, 0, 0, (LPCTSTR)textstring, strlen(textstring));
    BGI__ReleaseWinbgiDC( );
    RefreshWindow( NULL );
    // Todo: Change to refresh only the needed part.
}

// This function prints textstring to x,y
// POSTCONDITION: text has been written to the screen using the current font,
//                direction, and size. If a string is printed with the default
//                font using outtext or outtextxy, any part of the string that
//                extends outside the current viewport is truncated.
//
void outtextxy(int x, int y, char *textstring)
{
    HDC hDC = BGI__GetWinbgiDC( );
    WindowData* pWndData = BGI__GetWindowDataPtr( );

    // check alignment
    if (pWndData->alignment != TA_NOUPDATECP)
    {
	pWndData->alignment = TA_NOUPDATECP;
	set_align(pWndData);
    }

    TextOut(hDC, x, y, (LPCTSTR)textstring, strlen(textstring));
    BGI__ReleaseWinbgiDC( );

    RefreshWindow( NULL );
    // Todo: Change to refresh only the needed part.
}



// This function sets the vertical and horizontal justification based on CP
// POSTCONDITION: Text output is justified around the current position as
//                has been specified.
//
void settextjustify(int horiz, int vert)
{
    WindowData* pWndData = BGI__GetWindowDataPtr( );

    pWndData->textInfo.horiz = horiz;
    pWndData->textInfo.vert  = vert;

    WaitForSingleObject(pWndData->hDCMutex, 5000);
    set_align(pWndData);
    ReleaseMutex(pWndData->hDCMutex);
}


// This function sets the font and it's properties that will be used
// by all text output related functions.
// POSTCONDITION: text output after a call to settextstyle should be
//                in the newly specified format.
//
void settextstyle(int font, int direction, int charsize)
{
    WindowData* pWndData = BGI__GetWindowDataPtr( );

    pWndData->textInfo.font = font;
    pWndData->textInfo.direction = direction;
    pWndData->textInfo.charsize = charsize;

    WaitForSingleObject(pWndData->hDCMutex, 5000);
    set_font(pWndData);
    ReleaseMutex(pWndData->hDCMutex);
}

// This function sets the size of stroked fonts
// POSTCONDITION: these values will be used when charsize is zero in the
//                settextstyle assignments.  consequently output text will
//                be scaled by the appropriate x and y values when output.
//
void setusercharsize(int multx, int divx, int multy, int divy)
{
    WindowData* pWndData = BGI__GetWindowDataPtr( );
    pWndData->t_scale[0] = multx;
    pWndData->t_scale[1] = divx;
    pWndData->t_scale[2] = multy;
    pWndData->t_scale[3] = divy;

    WaitForSingleObject(pWndData->hDCMutex, 5000);
    set_font(pWndData);
    ReleaseMutex(pWndData->hDCMutex);
}

// This function returns the height in pixels of textstring using the current
// text output settings.
// POSTCONDITION: the height of the string in pixels has been returned.
//
int textheight(char *textstring)
{
    HDC hDC = BGI__GetWinbgiDC( );
    SIZE tb;

    GetTextExtentPoint32(hDC, (LPCTSTR) textstring, strlen(textstring), &tb);
    BGI__ReleaseWinbgiDC( );

    return tb.cy;
}

// This function returns the width in pixels of textstring using the current
// text output settings.
// POSTCONDITION: the width of the string in pixels has been returned.
//
int textwidth(char *textstring)
{
    HDC hDC = BGI__GetWinbgiDC( );
    SIZE tb;

    GetTextExtentPoint32(hDC, (LPCTSTR) textstring, strlen(textstring), &tb);
    BGI__ReleaseWinbgiDC( );

    return tb.cx;
}

void outstreamxy(int x, int y, std::ostringstream& out)
{
    std::string all, line;
    int i;
    int startx = x;

    all = out.str( );
    out.str("");

    moveto(x,y);
    for (i = 0; i < all.length( ); i++)
    {
	if (all[i] == '\n')
	{
	    if (line.length( ) > 0)
	    	outtext((char *) line.c_str( ));
	    y += textheight((char *) "X");
	    x = startx;
	    line.clear( );
	    moveto(x,y);
	}
	else
	    line += all[i];
    }
    if (line.length( ) > 0)
	outtext((char *) line.c_str( ));
}

void outstream(std::ostringstream& out)
{
    outstreamxy(getx( ), gety( ), out);
}



WinBGIm64/src/winthread.cxx
// File: WindowThread.cpp
// Written by:
//      Grant Macklem (Grant.Macklem@colorado.edu)
//      Gregory Schmelter (Gregory.Schmelter@colorado.edu)
//      Alan Schmidt (Alan.Schmidt@colorado.edu)
//      Ivan Stashak (Ivan.Stashak@colorado.edu)
// CSCI 4830/7818: API Programming
// University of Colorado at Boulder, Spring 2003
// http://www.cs.colorado.edu/~main/bgi
//

#include <windows.h>            // Provides the Win32 API
#include <windowsx.h>           // Provides message cracker macros (p. 96)
#include "winbgim.h"             // External API routines
#include "winbgitypes.h"        // Internal structures and routines
#include <vector>               // Used in BGI__WindowTable
#include <queue>                // Provides queue<POINTS>

// This structure is how the user interacts with the window.  Upon creation,y
// the user gets the index into this array of the window he created.  We will
// use this array for any function which does not deal with the window
// directly, but relies on the current window.
// MGM: hInstance is no longer a handle to the DLL, but instead it is
// a "self" handle, returned by GetCurrentThread( ).
std::vector<HWND> BGI__WindowTable;
int BGI__WindowCount = 0;                    // Number of windows currently in use
int BGI__CurrentWindow = NO_CURRENT_WINDOW;  // Index to current window
HINSTANCE BGI__hInstance;   // Handle to the instance of the DLL (creating the window class)

#include <iostream>
using namespace std;
// ID numbers for new options that are added to the system menu:
#define BGI_PRINT_SMALL 1
#define BGI_PRINT_MEDIUM 2
#define BGI_PRINT_LARGE 3
#define BGI_SAVE_AS 4
// This is the window creation and message processing function.  Each new
// window is created in its own thread and handles all its own messages.
// It creates the window, sets the current window of the application to this
// window, and signals the main thread that the window has been created.
//
DWORD WINAPI BGI__ThreadInitWindow( LPVOID pThreadData )
{
    HWND hWindow;                       // A handle to the window
    MSG Message;                        // A windows event message
    HDC hDC;                            // The device context of the window
    HBITMAP hBitmap;                    // A compatible bitmap of the DC for the Memory DC
    HMENU hMenu;                        // Handle to the system menu
    int CaptionHeight, xBorder, yBorder;
    
    WindowData *pWndData = (WindowData*)pThreadData; // Thread creation data

    if (pWndData->title.size( ))
    {
	CaptionHeight = GetSystemMetrics( SM_CYCAPTION );   // Height of caption
    }
    else
    {
	CaptionHeight = 0;                                 // Height of caption
    }
    xBorder = GetSystemMetrics( SM_CXFIXEDFRAME );      // Width of border
    yBorder = GetSystemMetrics( SM_CYFIXEDFRAME );      // Height of border
    
    int height = pWndData->height + CaptionHeight + 2*yBorder; // Calculate total height
    int width = pWndData->width + 2*xBorder;                   // Calculate total width

    int top = pWndData->inittop;                                 // MGM: Initial top
    int left = pWndData->initleft;                               // MGM: Initial left

    hWindow = CreateWindowEx( 0,                    // Extended window styles
                              _T( "BGILibrary" ),   // What kind of window
                              pWndData->title.c_str( ),  // Title at top of the window
			      pWndData->title.size( )
			      ? (WS_CAPTION | WS_SYSMENU | WS_MINIMIZEBOX | WS_DLGFRAME)
			      : (WS_POPUP|WS_DLGFRAME),
                              left, top, width, height,  // left top width height
                              NULL,                 // HANDLE to the parent window
                              NULL,                 // HANDLE to the menu
                              BGI__hInstance,       // HANDLE to this program
                              NULL );               // Address of window-creation data
    // Check if the window was created
    if ( hWindow == 0 )
    {
        showerrorbox( );
        return 0;
    }

    // Add the print options to the system menu, as shown in Chapter 10 (p 460) of Petzold
    hMenu = GetSystemMenu( hWindow, FALSE );
    AppendMenu( hMenu, MF_SEPARATOR, 0, NULL );
    AppendMenu( hMenu, MF_STRING, BGI_SAVE_AS, TEXT("Save image as...") );
    AppendMenu( hMenu, MF_STRING, BGI_PRINT_SMALL, TEXT("Print 2\" wide...") );
    AppendMenu( hMenu, MF_STRING, BGI_PRINT_MEDIUM, TEXT("Print 4.5\" wide...") );
    AppendMenu( hMenu, MF_STRING, BGI_PRINT_LARGE, TEXT("Print 7\" wide...") );
    AppendMenu( hMenu, MF_SEPARATOR, 0, NULL );

    // Store the HANDLE in the structure
    pWndData->hWnd = hWindow;

    // Store the address of the WindowData structure in the window's user data
    // MGM TODO: Change this to SetWindowLongPtr:
    //  SetWindowLongPtr( hWindow, GWL_USERDATA, (LONG_PTR)pWndData );
    SetWindowLongPtr( hWindow, GWLP_USERDATA, (LONG_PTR)pWndData );

    // Set the default active and visual page.  These must be set here in
    // addition to setting all the defaults in initwindow because the paint
    // method depends on using the correct page.
    pWndData->ActivePage = 0;
    pWndData->VisualPage = 0;

    // Clear the mouse handler array and turn off queuing
    memset( pWndData->mouse_handlers, 0, (WM_MOUSELAST-WM_MOUSEFIRST+1) * sizeof(Handler) );
    memset( pWndData->mouse_queuing, 0, (WM_MOUSELAST-WM_MOUSEFIRST+1) * sizeof(bool) );

    // Create a memory Device Context used for drawing.  The image is copied from here
    // to the screen in the paint method.  The DC and bitmaps are deleted
    // in cls_OnDestroy()
    hDC = GetDC( hWindow );
    pWndData->hDCMutex = CreateMutex(NULL, FALSE,	NULL);
    WaitForSingleObject(pWndData->hDCMutex, 5000);
    for ( int i = 0; i < MAX_PAGES; i++ )
    {
        pWndData->hDC[i] = CreateCompatibleDC( hDC );
        // Create a bitmap for the memory DC.  This is where the drawn image is stored.
        hBitmap = CreateCompatibleBitmap( hDC, pWndData->width, pWndData->height );
        pWndData->hOldBitmap[i] = (HBITMAP)SelectObject( pWndData->hDC[i], hBitmap );
    }
    ReleaseMutex(pWndData->hDCMutex);    
    // Release the original DC and set up the mutex for the hDC array
    ReleaseDC( hWindow, hDC );
    
    // Make the window visible
    ShowWindow( hWindow, SW_SHOWNORMAL );           // Make the window visible
    UpdateWindow( hWindow );                        // Flush output buffer
    
    // Tell the user thread that the window was created successfully
    SetEvent( pWndData->WindowCreated );

    // The message loop, which stops when a WM_QUIT message arrives
    while( GetMessage( &Message, NULL, 0, 0 ) )
    {
        TranslateMessage( &Message );
        DispatchMessage( &Message );
    }

    // Free memory used by thread structure
    delete pWndData;
    return (DWORD)Message.wParam;
}


// This function handles the WM_CHAR message.  This message is sent whenever
// the user presses a key in the window (after a WM_KEYDOWN and WM_KEYUP
// message, a WM_CHAR message is added).  It adds the key pressed to the
// keyboard queue for the window.
//
void cls_OnChar( HWND hWnd, TCHAR ch, int repeat )
{
    // This gets the address of the WindowData structure associated with the window
    WindowData *pWndData = BGI__GetWindowDataPtr( hWnd );

    pWndData->kbd_queue.push( (TCHAR)ch );// Add the key to the queue
    SetEvent( pWndData->key_waiting );    // Notify the waiting thread, if any
    FORWARD_WM_CHAR( hWnd, ch, repeat, DefWindowProc );
}

static void cls_OnClose( HWND hWnd )
{
    // This gets the address of the WindowData structure associated with the window
    WindowData *pWndData = BGI__GetWindowDataPtr( hWnd );

    exit(0);
}

// This function handles the destroy message.  It will cause the application
// to send WM_QUIT, which will then terminate the message pump thread.
//
static void cls_OnDestroy( HWND hWnd )
{
    // This gets the address of the WindowData structure associated with the window
    WindowData *pWndData = BGI__GetWindowDataPtr( hWnd );

    WaitForSingleObject(pWndData->hDCMutex, 5000);
    for ( int i = 0; i < MAX_PAGES; i++ )
    {
        // Delete the pen in the DC's
        DeletePen( SelectPen( pWndData->hDC[i], GetStockPen( WHITE_PEN ) ) );

        // Delete the brush in the DC's
        DeleteBrush( SelectBrush( pWndData->hDC[i], GetStockBrush( WHITE_BRUSH ) ) );

        // Here we clean up the memory device contexts used by the program.
        // This selects the original bitmap back into the memory DC.  The SelectObject
        // function returns the current bitmap which we then delete.
        DeleteObject( SelectObject( pWndData->hDC[i], pWndData->hOldBitmap[i] ) );
        // Finally, we delete the MemoryDC
        DeleteObject( pWndData->hDC[i] );
    }
    ReleaseMutex(pWndData->hDCMutex);
    // Clean up the bitmap memory
    DeleteBitmap( pWndData->hbitmap );

    // Delete the two events created
    CloseHandle( pWndData->key_waiting );
    CloseHandle( pWndData->WindowCreated );

    PostQuitMessage( 0 );
}


// This function handles the KEYDOWN message and will translate the virtual
// keys to the keys expected by the user
//
static void cls_OnKey( HWND hWnd, UINT vk, BOOL down, int repeat, UINT flags )
{
    // This gets the address of the WindowData structure associated with the window
    // TODO: Set event for each key
    WindowData *pWndData = BGI__GetWindowDataPtr( hWnd );

    switch ( vk )
    {
    case VK_CLEAR:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_CENTER );
        break;
    case VK_PRIOR:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_PGUP );
        break;
    case VK_NEXT:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_PGDN );
        break;
    case VK_END:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_END );
        break;
    case VK_HOME:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_HOME );
        break;
    case VK_LEFT:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_LEFT );
        SetEvent( pWndData->key_waiting );
        break;
    case VK_UP:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_UP );
        SetEvent( pWndData->key_waiting );
        break;
    case VK_RIGHT:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_RIGHT );
        SetEvent( pWndData->key_waiting );
        break;
    case VK_DOWN:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_DOWN );
        SetEvent( pWndData->key_waiting );
        break;
    case VK_INSERT:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_INSERT );
        break;
    case VK_DELETE:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_DELETE );
        break;
    case VK_F1:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F1 );
        break;
    case VK_F2:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F2 );
        break;
    case VK_F3:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F3 );
        break;
    case VK_F4:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F4 );
        break;
    case VK_F5:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F5 );
        break;
    case VK_F6:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F6 );
        break;
    case VK_F7:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F7 );
        break;
    case VK_F8:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F8 );
        break;
    case VK_F9:
        pWndData->kbd_queue.push( (TCHAR)0 );
        pWndData->kbd_queue.push( (TCHAR)KEY_F9 );
        break;
    }

    FORWARD_WM_KEYDOWN( hWnd, vk, repeat, flags, DefWindowProc );
}

#include <iostream>
static void cls_OnPaint( HWND hWnd )
{
    PAINTSTRUCT ps;             // BeginPaint puts info about the paint request here
    HDC hSrcDC;                 // The device context to copy from
    WindowData *pWndData = BGI__GetWindowDataPtr( hWnd );
    int width, height;          // Area that needs to be redrawn
    POINT srcCorner;            // Logical coordinates of the source image upper left point
    BOOL success;               // Is the BitBlt successful?
    int i;                      // Count for how many bitblts have been tried.

    WaitForSingleObject(pWndData->hDCMutex, INFINITE);
    BeginPaint( hWnd, &ps );
    hSrcDC = pWndData->hDC[pWndData->VisualPage];   // The source (memory) DC

    // Get the dimensions of the area that needs to be redrawn.
    width = ps.rcPaint.right - ps.rcPaint.left;
    height = ps.rcPaint.bottom - ps.rcPaint.top;
    

    // The region that needs to be updated is specified in device units (pixels) for the actual DC.
    // However, if a viewport is specified, the source image is referenced in logical
    // units.  Perform the conversion.
    // MGM TODO: When is the DPtoLP call needed?
    srcCorner.x = ps.rcPaint.left;
    srcCorner.y = ps.rcPaint.top;
    DPtoLP( hSrcDC, &srcCorner, 1 );

    // MGM: Screen BitBlts are not always successful, although I don't know why.
    success = BitBlt( ps.hdc, ps.rcPaint.left, ps.rcPaint.top, width, height,
			 hSrcDC, srcCorner.x, srcCorner.y, SRCCOPY );
    EndPaint( hWnd, &ps );  // Validates the rectangle
    ReleaseMutex(pWndData->hDCMutex);
    
    if ( !success )
    {   // I would like to invalidate the rectangle again
	// since BitBlt wasn't successful, but the recursion seems
	// to hang some machines.
	// delay(100);
	// InvalidateRect( hWnd, &(ps.rcPaint), FALSE );
	// std::cout << "Failure in cls_OnPaint" << std:: endl;
    }
}

// The message-handler function for the window
//
LRESULT CALLBACK WndProc
( HWND hWnd, UINT uiMessage, WPARAM wParam, LPARAM lParam )
{
    const std::queue<POINTS> EMPTY;
    POINTS where;
    
    WindowData *pWndData = BGI__GetWindowDataPtr( hWnd );
    int type;           // Type of mouse message
    Handler handler;    // Registered mouse handler
    UINT uHitTest;
    
    // If this is a mouse message, set our internal state
    if ( pWndData && ( uiMessage >= WM_MOUSEFIRST ) && ( uiMessage <= WM_MOUSELAST ) )
    {
	type = uiMessage - WM_MOUSEFIRST;
	if (!(pWndData->mouse_queuing[type]) && pWndData->clicks[type].size( ) )
	{
	    pWndData->clicks[type] = EMPTY;
	}
        pWndData->clicks[type].push(where = MAKEPOINTS( lParam ));  // Set the current position for the event type
	pWndData->mouse = where; // Set the current mouse position

        // If the user has registered a mouse handler, call it now
        handler = pWndData->mouse_handlers[type];
        if ( handler != NULL )
	    handler( where.x, where.y );
    }

    switch ( uiMessage )
    {
    HANDLE_MSG( hWnd, WM_CHAR, cls_OnChar );
    HANDLE_MSG( hWnd, WM_DESTROY, cls_OnDestroy );
    HANDLE_MSG( hWnd, WM_KEYDOWN, cls_OnKey );
    HANDLE_MSG( hWnd, WM_PAINT, cls_OnPaint );
    case WM_LBUTTONDBLCLK:
	return TRUE;
    case WM_NCHITTEST:
	uHitTest = DefWindowProc(hWnd, WM_NCHITTEST, wParam, lParam);
	if(uHitTest != HTCLIENT && pWndData && pWndData->title.size( ) == 0)
	    return HTCAPTION;
	else
	    return uHitTest;
    case WM_CLOSE:
	if ( pWndData->CloseBehavior )
	{
	    HANDLE_WM_CLOSE( hWnd, wParam, lParam, cls_OnClose );
	}
	return TRUE;
    case WM_SYSCOMMAND:
	switch ( LOWORD(wParam) )
	{
	case BGI_SAVE_AS: writeimagefile(NULL, 0, 0, INT_MAX, INT_MAX, false, hWnd); return 0;
	case BGI_PRINT_SMALL: printimage(NULL, 2.0, 0.75, 0.75, 0, 0, INT_MAX, INT_MAX, false, hWnd); return 0;
	case BGI_PRINT_MEDIUM: printimage(NULL, 4.5, 0.75, 0.75, 0, 0, INT_MAX, INT_MAX, false, hWnd); return 0;
	case BGI_PRINT_LARGE: printimage(NULL, 7.0, 0.75, 0.75, 0, 0, INT_MAX, INT_MAX, false, hWnd); return 0;
	}
	break;
    }
    return DefWindowProc( hWnd, uiMessage, wParam, lParam );
}

